### DISCUSSION TITLE
Does Nats has any generator which generates client code from schema just like grpc

### DISCUSSION THREAD

**Discussion by khanakia** [nats-io/nats.go#1153] [Q&A]
Do Nats have any client code generator from schema just like GRPC does so we can add Static Typing?

At the moment it's hard to know what parameters a message will accept so i want to generate a client which has Static Typing and make it easier to send the messages without even remembering what parameter a message will accept.

**Comment by bruth** [comment]
Hi @khanakia, messages in NATS are payload agnostic so you are free to choose your own schema strategy and generate types from an IDL, like protocol buffers, if you want. gRPC provides an extension on top of protocol buffers with the `service` type, however you can still separately generate `message` types and generate the types for your language of choice and serialize to/deserialize from the byte array a NATS message uses.

Hopefully that makes sense, but happy to show you an example if anything is unclear.

**Reply by khanakia** [reply]
Yes please, it would be better if see some example at least to get started with this :)

**Reply by bruth** [reply]
Sorry for the delay, here is a simple example: https://natsbyexample.com/examples/messaging/protobuf/go

**Reply by khanakia** [reply]
Thanks:
PS: I did not know we had this site https://natsbyexample.com/. It's a very useful site.

We should add the link to the site in the official docs (https://docs.nats.io/) also.

**Reply by chenjpu** [reply]
[nRPC is like gRPC, but over NATS](https://github.com/nats-rpc/nrpc)

**Reply by khanakia** [reply]
@chenjpu it seems that the repo is not maintained anymore. It hardly any new updates since 2020

**Reply by byebyebruce** [reply]
This is my work [https://github.com/byebyebruce/natsrpc](https://github.com/byebyebruce/natsrpc). Hope it can help you.
